### DISCUSSION TITLE
Advice needed for wireless I2C streaming

### DISCUSSION THREAD

**Discussion by leelx78** [pschatzmann/arduino-audio-tools#786] [Q&A] [answered]
Hi Phil (and everybody),

I'm completely blown away by this library, both in terms on quantity an quality. Amazing job!!

I'd like to use this for a project where I take analog audio, convert to I2S via a codec and then stream it via UDP on either WiFi or Ethernet but I wanted to have a quick chat with you guys on the way to do this the most efficient way so I don't waste time on something that has no chance of working properly. Obviously the people on this forum know a lot about audio, hence this message.

I'm reading I2S on ESP32-S2 (i.e. the audio source) using this line:

`esp_err_t result = i2s_read(I2S_PORT, &sBuffer, bufferLen, &bytesIn, portMAX_DELAY);`

which puts the data into the sBuffer, which is initialised as:
`#define bufferLen 1024
int16_t sBuffer[bufferLen];
`

..and write back the I2S into the DAC (i.e. the receiver) with this:
`esp_err_t result_w = i2s_write(I2S_PORT, &sBuffer, bytesIn, &bytesOut, portMAX_DELAY);`

If I leave it at this, I could in principle send sBuffer to UDP and handle all data that way but it would use too much bandwidth (~1.4Mbps for 44.1kHz/16bit/stereo) so I was thinking of using an LC3 encoder/decoder and I came across this example (https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/tests/codecs/test-codec-lc3/test-codec-lc3.ino) using your library.

What's the most efficient way to use the LC3 encoder and decoder with my data structure of simple int16_t arrays? Most of your examples are a lot fancier than this so I wanted to make sure I wasn't overcomplicating things.

Any other pointers? I'm aiming at keeping latency around 100ms, so I would need encoders and decoders that can be fast and efficient on ESP32-S2+

Let me know what you think. Any suggestions are very welcome :)

thanks a lot!

**Comment by pschatzmann** [comment, marked as answer]
I suggest to keep with the philosophy and just and just use a StreamCopy  to copy an I2SStream to a standard WiFiUDP object (which is also Stream).  You can also use by UDPStream class which simplifies the processing a bit.

You will find examples how to use encoders and decoders and they should work with this scenario as well.

However be careful, using UDP you might loose data, which leads to some hickups when you use PCM data, but which is a major disaster when you use encoded data.

**Reply by leelx78** [reply]
thanks Phil, 

the only issue with I2SStream is that I was planning to read the data into int16_t using i2s_read and i2s_write which also included all the ESP32 I2S configuration and such. Do I NEED to use I2SStream to read I2S or can I copy raw int16_t data directly to UDPStream?

A 1-2 line example would greatly help clarifying here if possible.

Sorry if this is a trivial question :)

thanks!

**Reply by pschatzmann** [reply]
Sure you can, but there is no reason to do so and you don't gain anything. On the contrary your sketch will just get longer.
You could also use the I2S API provided by ESP32 Arduino which is also stream based.

**Reply by leelx78** [reply]
thanks, so this is my loop:
```
void loop()
{
  // Get I2S data and place in data buffer
  size_t bytesIn = 0;
  size_t bytesOut = 0;
  esp_err_t result = i2s_read(I2S_PORT, &sBuffer, bufferLen, &bytesIn, portMAX_DELAY);

  if (result == ESP_OK)
  {
    // Send what we just received back to the codec
    esp_err_t result_w = i2s_write(I2S_PORT, &sBuffer, bytesIn, &bytesOut, portMAX_DELAY);

    // If there was an I2S write error, let us know on the serial terminal
    if (result_w != ESP_OK)
    {
      Serial.print("I2S write error.");
    }
  }
  ```
  
  which is working well in a passthrough (Audio -> ADC -> I2S -> DAC -> Headphones). The original idea was to 
  - read 1024 bytes
  - compress the packet to reduce size
  - send it out via whatever method (I would handle the stream directly)
  - the receiver then decompresses it and adds it to the I2S queue at the receiver end via i2s_write.
  
I guess the issue with this is that compressing the packet would not result in much lower data rate but it has the advantage of each packet being self contained and less susceptible to noise if one packet is missed. Is this the case? Have you tried something of this type?
  
Your approach is more stream oriented which looks like a more professional way of dealing with this so I was looking at this line:
  `EncodedAudioStream encoder(&stream, new LC3Encoder()); // encode
`
but the issue I have right now is that it's not clear to me is:
- how to create the "stream" input object given I have int16_t data
- how to poll int16_t data from the encoder so that I can send it via UDP

Would you have a minute to sketch down 1-2 Iines of code to show that? This is not me being lazy, it's just for clarification since I'm really not familiar with the stream approach in Arduino and so far I can't see how I can use this within my code.

thanks again for the help!

**Comment by pschatzmann** [comment]
But all the data used via streams is usually int16_t unless you you define a bits_per_sample other then 16! In the API it's just defined as byte array so that we can process any type of data!

**Reply by leelx78** [reply]
wait, maybe this then?
`AudioStream stream;
EncodedAudioStream encoder(&stream, new LC3Encoder());
`

and in the loop I need to keep calling:
`
stream.write(sBuffer, bufferLen);
`
correct?

**Comment by pschatzmann** [comment]
In the EncodedAudioStream you need to specify the output stream! As explained above this can be a WiFiUDP or a UDPStream object.
In the loop you need to check how much you have sent and reprocess the missing bytes!

That's the reason why I told you you should use a StreamCopy which will automatically take care of this...

**Reply by leelx78** [reply]
nevermind

**Reply by pschatzmann** [reply]
I suggest to read the Wiki: it is explained there!

**Reply by leelx78** [reply]
thanks Phil, I finally had enough time to go through your approach in more detail and it's working now. 

I have one last question but I'll start a new thread for that.
